### PR TITLE
Implement basic course and instructor search (excludes sorting)

### PR DESCRIPTION
**Pull request by shannon-heh**
# Goals
- Add basic api/backend logic to search for courses & instructors
- Basically re-implemented Princeton Courses search, but with some changes (check out the assumptions I wrote in the code)

# Future Work
- Work with frontend to decide how query input and results should be formatted
- Sanitizing inputs (if needed)
- Return course information for retrieved instructors
- Add sorting for queried results

### CODE REVIEW

**Comment by shannon-heh** [comment]
@jkuo168 @howard-yen This PR is for the search endpoint. I still need to implement sorting + do more testing + change the return type, but the basic search logic is already here if you wanted to get started on the frontend. To understand how the query input is passed in,  read the comments in `database_api.py`.

**Comment by jkuo168** [comment]
There are no bugs for this endpoint atm right? Just not fully completed? If so, can you merge to main, so I can pull. I would merge this branch to my branch, but it might cause future merging issues bc I will have these changes, but your finished request would have these changes but also different changes as well.

**Comment by shannon-heh** [comment]
@jkuo168 i want to fix one small bug and i'll merge sometime today (sorry i'm busy packing so i might not get to merging until later tonight)

**Review by jkuo168 — APPROVED** [review]

File: database_api.py [hunk 1]
```diff
@@ -50,9 +51,317 @@ def get_all_test(self):
...
     db = DatabaseAPI()
+
+    """
+        query = {
+            "string": "the query string",
+            "semester": "one of the semester",
+            "sort": "Course Code, Relevance, Rating, Title",
+            "filters": {
+                "special": "*" or "NEW",
+                "dist": [list of dist codes],
```
Comment (jkuo168) [inline database_api.py:250]: dist or dists?

File: database_api.py [hunk 2]
```diff
@@ -50,9 +51,317 @@ def get_all_test(self):
...
+                "grading": [list of grading categorise],
+                "level": [list of levels],
```
Comment (jkuo168) [inline database_api.py:252]: level or levels?

File: app.py [hunk 3]
```diff
@@ -62,3 +62,15 @@ def path():
 def test_database():
     logger.info("Reached /api/test-database")
     return jsonify(db.get_all_test())
+
+@app.route("/search", methods=["GET", "POST"])
+def search():
+    if request.method == "GET":
+        query = request.args.get("query")
```
Comment (jkuo168) [inline app.py:69]: this returns a string. to convert import json and do json.loads(request.args.get("query")). this should convert it to a json object

File: database_api.py (hunk 1 above)
Comment (shannon-heh) [reply database_api.py:250]: dists, sorry i changed the key names in between

File: database_api.py (hunk 2 above)
Comment (shannon-heh) [reply database_api.py:252]: levels

**Comment by shannon-heh** [comment]
@jkuo168
changes made in my most recent commit (https://github.com/PrincetonResInDe/course-selection/pull/19/commits/f74cc3e52290f4ba6b5f5ff438dec7713d4c8ee4)
- `search` endpoint now only accepts GET requesets
- input args to `search` endpoint are now separate strings
- `search` endpoint now returns valid response: a jsonified tuple, the first value of the tuple being a list of courses, the second value being a list of instructors
- `search` logic has been refactored into `search.py`

**Review by jkuo168 — APPROVED** [review]
lgtm!

**Pull request merged**